ipytone/ipytone.py

```diff
@@ -4,26 +4,73 @@
 # Copyright (c) Benoit Bovy.
 # Distributed under the terms of the Modified BSD License.
 
-from ipywidgets import Widget
-from traitlets import Unicode, Float, Int, Bool, validate, TraitError
+from ipywidgets import widget_serialization, Widget
+from traitlets import Instance, Union, Unicode, List, Float, Int, Bool, validate, TraitError
+
 from ._frontend import module_name, module_version
 
 
-class Oscillator(Widget):
+class _ToneWidgetBase(Widget):
+    _model_module = Unicode(module_name).tag(sync=True)
+    _model_module_version = Unicode(module_version).tag(sync=True)
+
+
+class AudioNode(_ToneWidgetBase):
+    """An audio node widget."""
+
+    _model_name = Unicode('AudioNodeModel').tag(sync=True)
+
+    inputs = List(Instance(Widget)).tag(sync=True, **widget_serialization)
+    outputs = List(Instance(Widget)).tag(sync=True, **widget_serialization)
+
+
+class Destination(AudioNode):
+    """Audio master output node."""
+
+    _singleton = None
+
+    _model_name = Unicode('DestinationModel').tag(sync=True)
+
+    mute = Bool(False).tag(sync=True)
+    volume = Float(-16).tag(sync=True)
+
+    def __new__(cls):
+        if Destination._singleton is None:
+            Destination._singleton = super(Destination, cls).__new__(cls)
+        return Destination._singleton
+
+
+_DESTINATION = Destination()
+
+
+def get_destination():
+    """Returns ipytone's audio master output node."""
+    return _DESTINATION
+
+
+class Oscillator(AudioNode):
     """A simple Oscillator."""
    
     _model_name = Unicode('OscillatorModel').tag(sync=True)
-    _model_module = Unicode(module_name).tag(sync=True)
-    _model_module_version = Unicode(module_version).tag(sync=True)
 
-    type = Unicode("sine", help="Oscillator type").tag(sync=True)
-    frequency = Float(440, help="Oscillator frequency").tag(sync=True)
-    detune = Int(0, help="Oscillator frequency detune").tag(sync=True)
-    volume = Float(-16, help="Oscillator gain").tag(sync=True)
-    started = Bool(False, help="Start/stop oscillator").tag(sync=True)
+    type = Unicode("sine").tag(sync=True)
+    frequency = Float(440).tag(sync=True)
+    detune = Int(0).tag(sync=True)
+    volume = Float(-16).tag(sync=True)
+    started = Bool(False).tag(sync=True)
+
+    def __init__(self, type='sine', frequency=440, detune=0, volume=-16, started=False, **kwargs):
+        super(Oscillator, self).__init__(
+            type=type,
+            frequency=frequency,
+            detune=detune,
+            volume=volume,
+            started=started,
+            **kwargs
+        )
```

Comment:
I'm not sure why you propagate all the attributes to the base class :thinking:

Reply:
I'm not sure either 🙂. This is very much work in progress..

Reply:
OK, no worries, this PR looks great!